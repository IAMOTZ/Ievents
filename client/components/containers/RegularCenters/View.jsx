import React from 'react';
import PropTypes from 'prop-types';
import { RegularTopNavigation } from '../../common/TopNavigation';
import Footer from '../../common/Footer';
import CenterCards from '../../common/CenterCards';
import { LoadingBox } from '../../common/LoadingAnimation';
import { CenterDetailsModal } from '../../common/Modals';
import Pagination from '../../common/Pagination';

const View = (props) => {
  const renderAlternateBody = () => {
    if (props.centers.length === 0 && props.fetchingCenterStarted) {
      return <LoadingBox iconSize={4} />;
    } else if (props.centers.length === 0 && !props.fetchingCenterStarted) {
      return (
        <div className="page-content empty-centers-list text-center">
          <h1 className="display-3">The centers list is empty</h1>
          <h3 className="font-weight-normal">Check back in a moment...</h3>
        </div>
      );
    } return null;
  };
  return (
    <div id="regular-centers-container">
      <RegularTopNavigation />
      {
        renderAlternateBody() ||
          <div className="page-content mb-5">
            <div className="container">
              <h1 className="caption-text">A special center for a special event</h1>
              {/* Centers Grid */}
              <div className="row">
                <CenterCards
                  centers={props.centers}
                  onBook={props.onBook}
                  createModalContent={props.createModalContent}
                />
              </div>
            </div>
            <Pagination
              pageCount={Math.ceil(props.pagination.totalCount / props.pagination.limit)}
              onPageChange={props.updatePagination}
            />
            <CenterDetailsModal {...props.modalContent} />
          </div>
      }
      <Footer />
    </div>
  );
};

View.defaultProps = {
  modalContent: {}
};

/* eslint-disable react/forbid-prop-types */
View.propTypes = {
  centers: PropTypes.array.isRequired,
  onBook: PropTypes.func.isRequired,
  fetchingCenterStarted: PropTypes.bool.isRequired,
  pagination: PropTypes.object.isRequired,
  updatePagination: PropTypes.func.isRequired,
  modalContent: PropTypes.object,
  createModalContent: PropTypes.func.isRequired
};

export default View;
