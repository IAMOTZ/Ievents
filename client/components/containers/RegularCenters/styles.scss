@import "../../../styles/variables";

#regular-centers-container {
    .caption-text {
      font-size: 1.5rem;
      font-weight: 300;
      line-height: 1.2;
      text-align: center;
      margin-bottom: 1.5rem;
    }
    .empty-centers-list {
     padding-top: 75px;
     min-height: 75vh;
    }
    @include media-breakpoint-up("sm") {
      .caption-text {
        padding-top: 70px;
        font-size: 3.5rem;
      }
    }
}