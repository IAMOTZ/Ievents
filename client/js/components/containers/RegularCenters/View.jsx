import React from 'react';
import PropTypes from 'prop-types';
import { RegularTopNavigation } from '../../common/TopNavigation';
import Footer from '../../common/Footer';
import CenterCards from '../../common/CenterCards';
import { CenterModal } from '../../common/Modals';
import { LoadingBox } from '../../common/LoadingAnimation';

const View = props => (
  <div id="centers1-container">
    <RegularTopNavigation />
    {
      props.centers.length === 0 && props.fetchingCenterStarted ?
        <LoadingBox iconSize={4} /> :
        <section id="centers-section" className="mb-5">
          <div className="container">
            {/* Search Box */}
            <div
              id="center-section-content"
              className="d-flex flex-column align-items-center"
            >
              <div id="center-section-header">
                <h1 className="display-4">A special center for a special event</h1>
              </div>
              <div className="search-box input-group w-50 mt-5">
                <input
                  type="text"
                  className="form-control"
                  placeholder="Search for a center"
                />
                <div className="input-group-btn d-none d-sm-block">
                  <button
                    type="button"
                    className="btn btn-secondary dropdown-toggle"
                    data-toggle="dropdown"
                  >Search with
                  </button>
                  <div className="dropdown-menu dropdown-menu-right">
                    <a className="dropdown-item" href="#">Location</a>
                    <div role="separator" className="dropdown-divider" />
                    <a className="dropdown-item" href="#">Name</a>
                  </div>
                </div>
              </div>
              <div className="search-box mt-3 d-block d-sm-none">
                <button
                  type="button"
                  className="btn btn-secondary dropdown-toggle"
                  data-toggle="dropdown"
                >Search with
                </button>
                <div className="dropdown-menu">
                  <a className="dropdown-item" href="#">Location</a>
                  <div role="separator" className="dropdown-divider" />
                  <a className="dropdown-item" href="#">Name</a>
                </div>
              </div>
            </div>
            {/* Centers Grid */}
            <div className="mt-5">
              <div className="row">
                <CenterCards centers={props.centers} btnAction={props.showModal} />
              </div>
            </div>
          </div>
        </section>
    }
    {/* Modal */}
    {
      props.modalContent ? <CenterModal
        modalContent={props.modalContent}
        onBook={props.onBook}
        redirectPath="/users/login"
      /> : null
    }
    <Footer />
  </div>
);

View.defaultProps = {
  modalContent: {},
};

/* eslint-disable react/forbid-prop-types */
View.propTypes = {
  centers: PropTypes.array.isRequired,
  onBook: PropTypes.func.isRequired,
  showModal: PropTypes.func.isRequired,
  modalContent: PropTypes.object,
  fetchingCenterStarted: PropTypes.bool.isRequired,
};

export default View;

