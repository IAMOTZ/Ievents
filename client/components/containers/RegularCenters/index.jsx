import React from 'react';
import PropTypes from 'prop-types';
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { find } from 'lodash';
import { getAllCenters } from '../../../actions/centerActions';
import { setCenterToBook } from '../../../actions/eventActions';
import './styles.scss';
import View from './View';

@connect(store => (
  {
    isUserAuthenticaed: store.authReducer.loggingUserResolved,
    fetchingCenterStarted: store.fetchCentersReducer.fetchingCenterStarted,
    centers: store.fetchCentersReducer.centers,
  }
))
class RegularCenters extends React.Component {
  constructor() {
    super();
    this.state = {
      modalContent: null,
    };
  }
  componentWillMount() {
    this.props.dispatch(getAllCenters());
  }

  /**
   * It updates the store about a center that is to be booked.
   * @param {Number} centerId The ID of the center.
   */
  onBook = (centerId) => {
    this.props.dispatch(setCenterToBook(centerId));
  }

  /**
   * Displays the center modal.
   * It uses the center ID to get the details of the center to show.
   * @param {Event} event The event object.
   */
  showModal = (event) => {
    const state = { ...this.state };
    const centerId = Number(event.target.id);
    state.modalContent = find(this.props.centers, { id: centerId });
    this.setState(state);
  }

  render() {
    if (this.props.isUserAuthenticaed) {
      return <Redirect to="/events" />;
    }
    return (
      <View
        centers={this.props.centers}
        onBook={this.onBook}
        showModal={this.showModal}
        modalContent={this.state.modalContent}
        fetchingCenterStarted={this.props.fetchingCenterStarted}
      />
    );
  }
}

RegularCenters.defaultProps = {
  isUserAuthenticaed: false,
  fetchingCenterStarted: false,
  centers: [],
  dispatch: () => {},
};

/* eslint-disable react/forbid-prop-types */
RegularCenters.propTypes = {
  isUserAuthenticaed: PropTypes.bool,
  fetchingCenterStarted: PropTypes.bool,
  centers: PropTypes.array,
  dispatch: PropTypes.func,
};

export default RegularCenters;
