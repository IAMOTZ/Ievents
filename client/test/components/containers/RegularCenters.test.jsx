import React from 'react';
import RegularCenters from '../../../components/containers/RegularCenters';
import RegularCentersView from '../../../components/containers/RegularCenters/View';

const sampleCenter = {
  id: 1,
  name: 'test-center1',
  location: 'test-center-location',
  details: 'test-center-details',
  capacity: '200',
  price: '500',
  images: ['http://testImage.jpg'],
};

describe('<RegularCenters />', () => {
  const props = {
    isUserAuthenticaed: false,
    fetchingCenterStarted: false,
    centers: [sampleCenter],
    pagination: {},
    dispatch: jest.fn(),
  };
  const alterProps = newProps => ({ ...props, ...newProps });
  describe('Rendering:', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<RegularCenters {...props} />);
    });
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('should redirect to centers page if user is authenticated', () => {
      wrapper.setProps(alterProps({ isUserAuthenticaed: true }));
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('Behaviour:', () => {
    let wrapper; let wrapperInstance;
    beforeEach(() => {
      props.dispatch.mockReset();
      wrapper = shallow(<RegularCenters {...props} />);
      wrapperInstance = wrapper.instance();
    });
    // The names on the following test block refers
    // to the instance method that handles the behaviour tested.
    describe('COMOPONENT_WILL_MOUNT', () => {
      it('should dispatch an action', () => {
        expect(props.dispatch.mock.calls.length).toEqual(1);
      });
    });

    describe('ON_BOOK', () => {
      it('should dispatch an action', () => {
        props.dispatch.mockReset();
        wrapperInstance.onBook(sampleCenter.id);
        expect(props.dispatch.mock.calls.length).toEqual(1);
      });
    });

    describe('CREATE_MODAL_CONTENT', () => {
      it('should update the state with the content of a modal', () => {
        wrapperInstance.createModalContent(sampleCenter);
        expect(wrapper.state('modalContent')).toEqual({
          centerName: sampleCenter.name,
          centerImages: sampleCenter.images,
          centerDetails: sampleCenter.details
        });
      });
    });

    describe('UPDATE_PAGINATION', () => {
      it('should dispatch an action', () => {
        props.dispatch.mockReset();
        const fakePageData = { selected: 3 };
        wrapperInstance.updatePagination(fakePageData);
        expect(props.dispatch.mock.calls.length).toEqual(1);
      });
    });
  });
});

describe('<RegularCentersView />', () => {
  const props = {
    centers: [],
    onBook: jest.fn(),
    fetchingCenterStarted: false,
    pagination: {},
    updatePagination: jest.fn(),
    modalContent: {},
    createModalContent: jest.fn(),
  };
  const alterProps = newProps => ({ ...props, ...newProps });
  describe('Rendering', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<RegularCentersView {...props} />);
    });
    it('should render correctly when there are no centers and it is fetching', () => {
      wrapper.setProps(alterProps({ fetchingCenterStarted: true }));
      expect(wrapper).toMatchSnapshot();
    });
    it('should render correctly when there are centers and it is not fetching', () => {
      wrapper.setProps(alterProps({ centers: [sampleCenter] }));
      expect(wrapper).toMatchSnapshot();
    });
  });
});
