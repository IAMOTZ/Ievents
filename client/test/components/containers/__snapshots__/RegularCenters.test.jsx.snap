// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegularCenters /> Rendering: should redirect to centers page if user is authenticated 1`] = `
<Redirect
  push={false}
  to="/centers"
/>
`;

exports[`<RegularCenters /> Rendering: should render correctly 1`] = `
<View
  centers={
    Array [
      Object {
        "capacity": "200",
        "details": "test-center-details",
        "id": 1,
        "images": Array [
          "http://testImage.jpg",
        ],
        "location": "test-center-location",
        "name": "test-center1",
        "price": "500",
      },
    ]
  }
  createModalContent={[Function]}
  fetchingCenterStarted={false}
  modalContent={null}
  onBook={[Function]}
  pagination={Object {}}
  updatePagination={[Function]}
/>
`;

exports[`<RegularCentersView /> Rendering should render correctly when there are centers and it is not fetching 1`] = `
<div
  id="regular-centers-container"
>
  <RegularTopNavigation />
  <div
    className="page-content mb-5"
  >
    <div
      className="container"
    >
      <h1
        className="caption-text"
      >
        A special center for a special event
      </h1>
      <div
        className="row"
      >
        <CenterCards
          centers={
            Array [
              Object {
                "capacity": "200",
                "details": "test-center-details",
                "id": 1,
                "images": Array [
                  "http://testImage.jpg",
                ],
                "location": "test-center-location",
                "name": "test-center1",
                "price": "500",
              },
            ]
          }
          createModalContent={[MockFunction]}
          isAdmin={false}
          isTransactionsPage={false}
          onBook={[MockFunction]}
          onEdit={[Function]}
          onViewTransactions={[Function]}
        />
      </div>
    </div>
    <Pagination
      onPageChange={[MockFunction]}
      pageCount={NaN}
    />
    <CenterDetailsModal
      centerDetails=""
      centerImages={
        Array [
          "/images/defaultImgx4.jpeg",
        ]
      }
      centerName=""
    />
  </div>
  <Footer />
</div>
`;

exports[`<RegularCentersView /> Rendering should render correctly when there are no centers and it is fetching 1`] = `
<div
  id="regular-centers-container"
>
  <RegularTopNavigation />
  <LoadingBox
    iconSize={4}
  />
  <Footer />
</div>
`;
